Mathlib/RingTheory/ZMod/UnitsCyclic.lean

```diff
@@ -0,0 +1,369 @@
+/-
+Copyright (c) 2025 Antoine Chambert-Loir. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Antoine Chambert-Loir
+-/
+import Mathlib.Analysis.Normed.Ring.Lemmas
+import Mathlib.Data.Nat.Choose.Dvd
+import Mathlib.Data.ZMod.Units
+import Mathlib.Data.ZMod.Zero
+import Mathlib.GroupTheory.SpecificGroups.ZGroup
+import Mathlib.Algebra.Order.Star.Basic
+
+/-! # Cyclicity of the units of `ZMod n`
+
+`ZMod.isCyclic_units_iff` : `(ZMod n)ˣ` is cyclic iff
+one of the following mutually exclusive cases happens:
+  - `n = 0` (then `ZMod 0 ≃+* ℤ` and the group of units is cyclic of order 2);
+  - `n = `1`,  `2`  or `4`
+  - `n` is a power `p ^ e` of an odd prime number, or twice such a power
+  (with `1 ≤ e`).
+
+The individual cases are proved by `inferInstance` and are
+also directly provided by :
+
+* `ZMod.isCyclic_units_zero`
+* `ZMod.isCyclic_units_one`
+* `ZMod.isCyclic_units_two`
+* `ZMod.isCyclic_units_four`
+
+The case of prime numbers is also an instance:
+
+* `ZMod.isCyclic_units_prime`
+
+* `ZMod.not_isCyclic_units_eight`: `(ZMod 8)ˣ` is not cyclic
+
+* `ZMod.orderOf_one_add_mul_prime`: the order of `1 + a * p`
+modulo `p ^ (n + 1)` is `p ^ n` when `p` does not divide `a`.
+
+* `ZMod.orderOf_five` : the order of `5` modulo `2 ^ (n + 3)` is `2 ^ (n + 1)`.
+
+* `ZMod.isCyclic_units_of_prime_pow` : the case of odd prime powers
+
+* `ZMod.isCyclic_units_two_pow_iff` : `(ZMod (2 ^ n))ˣ` is cyclic iff `n ≤ 2`.
+
+The proofs follow [Ireland and Rosen,
+  *A classical introduction to modern number theory*, chapter 4]
+  [IrelandRosen1990]
+but build on a general induction divisibility lemma suggested by Junyan Xu.
+
+-/
+
+open scoped Nat
+
+namespace ZMod
+
+section EasyCases
+
+theorem isCyclic_units_zero :
+    IsCyclic (ZMod 0)ˣ := inferInstance
+
+theorem isCyclic_units_one :
+    IsCyclic (ZMod 1)ˣ := inferInstance
+
+theorem isCyclic_units_two :
+    IsCyclic (ZMod 2)ˣ := inferInstance
+
+theorem isCyclic_units_four :
+    IsCyclic (ZMod 4)ˣ := by
+  apply isCyclic_of_prime_card (p := 2)
+  simp only [Nat.card_eq_fintype_card, card_units_eq_totient]
+  rfl
+
+/- The multiplicative group of `ZMod p` is cyclic. -/
+theorem isCyclic_units_prime {p : ℕ} (hp : p.Prime) :
+    IsCyclic (ZMod p)ˣ :=
+  have : Fact (p.Prime) := ⟨hp⟩
+  inferInstance
+
+theorem not_isCyclic_units_eight :
+    ¬ IsCyclic (ZMod 8)ˣ := by
+  rw [IsCyclic.iff_exponent_eq_card, Nat.card_eq_fintype_card, card_units_eq_totient]
+  have h : Monoid.exponent (ZMod 8)ˣ ∣ 2 := Monoid.exponent_dvd_of_forall_pow_eq_one (by decide)
+  intro (h' : Monoid.exponent (ZMod 8)ˣ = 4)
+  simp [h'] at h
+
+end EasyCases
+
+section Divisibility
+
+variable {R : Type*} [CommSemiring R] {u v : R} {p : ℕ}
+
+/- This lemma is due to Junyan Xu -/
```

Comment:
```suggestion
-- This lemma is due to Junyan Xu
```
I don't think we keep track of authors of lemmas (nothing personal against Junyan of course, it would be a total mess to actually do it). Maybe you can add him to the list of authors of the file?